Source/ocdm/open_cdm_impl.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * If not stated otherwise in this file or this component's LICENSE file the
+ * following copyright and licenses apply:
+ *
+ * Copyright 2020 Metrological
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Module.h"
+#include "open_cdm_impl.h"
+
+class SessionPrivate {
+    private:
+        typedef uint32_t (*ConstructSessionPrivate)(struct OpenCDMSession*, void*&);
+        typedef uint32_t (*DestructSessionPrivate)(struct OpenCDMSession*, void*&);
+
+    public:
+        SessionPrivate(const SessionPrivate&) = delete;
+        SessionPrivate& operator= (const SessionPrivate&) = delete;
+
+        SessionPrivate()
+            : _constructSessionPvt(nullptr)
+            , _destructSessionPvt(nullptr) {
+                Load();
+        }
+        ~SessionPrivate() = default;
+
+        OpenCDMError Construct(struct OpenCDMSession* session, void* &pvtData)
+        {
+            OpenCDMError result = OpenCDMError::ERROR_METHOD_NOT_IMPLEMENTED;
+            if(_constructSessionPvt != nullptr) {
+                result = ( _constructSessionPvt(session, pvtData) == 0 ? OpenCDMError::ERROR_NONE : OpenCDMError::ERROR_UNKNOWN);
+            }
```

Comment:
if there would be an else here where the pbtData is set to null, one could live without having the:
EXTERNAL uint32_t opencdm_construct_session_private(struct OpenCDMSession* session, void* &pvtData);
or 
EXTERNAL uint32_t opencdm_destruct_session_private(struct OpenCDMSession* session, void* &pvtData); 
methods in the adapter. So only for the Realtek/Amlogic adapters you need to add those, all other variant can stay as they where.
If we would implement them we could just link against them at compile time as we would expect them than. It ould make the code faster but would be not backwards compatible.... It is what you prefer... Lets discuss on skype later ...

Reply:
Updated.